pkg/scheduler/metrics/metric_recorder.go

```diff
@@ -186,39 +200,42 @@ func (r *MetricAsyncRecorder) observeMetricAsync(m *metrics.HistogramVec, value
 	}
 	select {
 	case r.bufferCh <- newMetric:
+		r.signalFlush()
 	default:
 	}
 }
 
 // run flushes buffered metrics into Prometheus every second.
 func (r *MetricAsyncRecorder) run() {
+	timer := time.NewTimer(r.interval)
+
 	for {
 		select {
 		case <-r.stopCh:
 			close(r.IsStoppedCh)
 			return
-		default:
+		case <-timer.C:
+		case <-r.flushCh:
 		}
+
 		r.FlushMetrics()
-		time.Sleep(r.interval)
+
+		// We don’t want any extra flushes.
+		// Within each interval, there should be exactly one flush triggered by the timer.
+		timer.Reset(r.interval)
```

Comment:
`timer.C` is not drained in this patch, because after go 1.23, I believe that it's not necessary anymore.
correct me if I'm wrong, or there is something i missed.

```
// For a chan-based timer created with NewTimer, as of Go 1.23,
// any receive from t.C after Reset has returned is guaranteed not
// to receive a time value corresponding to the previous timer settings;
// if the program has not received from t.C already and the timer is
// running, Reset is guaranteed to return true.
// Before Go 1.23, the only safe way to use Reset was to call [Timer.Stop]
// and explicitly drain the timer first.
// See the [NewTimer] documentation for more details.
func (t *Timer) Reset(d Duration) bool {
```